### PR TITLE
Replace the busy wait from AtomicSafeInitializer with an alternate method that eliminates wasted CPU cycles

### PR DESCRIPTION
**Pull request by artnaseef**
This patch eliminates the busy-wait in AtomicSafeInitializer that consumes significant CPU cycles (confirmed via test program) for each thread waiting for the initialized result while the initializer runs.

### CODE REVIEW

**Comment by artnaseef** [comment]
I created a test program that verifies the original code is performing a busy wait; I'll gladly share it.  It may be possible to write a unit test that checks for that as well.

**Comment by artnaseef** [comment]
I'm going to post an alternative that's less of a change and avoids the use of synchronized blocks and notifications.

File: src/main/java/org/apache/commons/lang3/concurrent/AtomicSafeInitializer.java [hunk 1]
```diff
@@ -73,12 +72,31 @@
     public final T get() throws ConcurrentException {
         T result;
 
-        while ((result = reference.get()) == null) {
-            if (factory.compareAndSet(null, this)) {
-                reference.set(initialize());
+        synchronized ( reference ) {
```
Comment (ecki) [inline src/main/java/org/apache/commons/lang3/concurrent/AtomicSafeInitializer.java]: I dont think this is a good idea to make a lock free code use such a large synchronized block now.

**Comment by ecki** [comment]
Ah sorry, missed your comment that you re-implement it. :)

**Pull request closed without merging**